toolchain/check/type.cpp

```diff
@@ -238,32 +238,28 @@ auto GetUnboundElementType(Context& context, SemIR::TypeInstId class_type_id,
                                                 element_type_id);
 }
 
-auto GetCanonicalizedFacetOrTypeValue(Context& context, SemIR::InstId inst_id)
+auto GetCanonicalFacetOrTypeValue(Context& context, SemIR::InstId inst_id)
     -> SemIR::InstId {
-  // We can have FacetAccessType of a FacetValue, and a FacetValue of a
-  // FacetAccessType, but they don't nest indefinitely.
-  if (auto access = context.insts().TryGetAs<SemIR::FacetAccessType>(inst_id)) {
-    inst_id = access->facet_value_inst_id;
+  if (inst_id == SemIR::ErrorInst::InstId) {
+    return inst_id;
   }
-  if (auto value = context.insts().TryGetAs<SemIR::FacetValue>(inst_id)) {
-    inst_id = value->type_inst_id;
 
-    if (auto access =
-            context.insts().TryGetAs<SemIR::FacetAccessType>(inst_id)) {
-      inst_id = access->facet_value_inst_id;
-    }
-  }
+  CARBON_CHECK(
+      context.types().IsFacetType(context.insts().Get(inst_id).type_id()),
+      "{0}", context.insts().Get(inst_id).type_id());
 
-  CARBON_CHECK(!context.insts().Is<SemIR::FacetAccessType>(inst_id));
-  CARBON_CHECK(!context.insts().Is<SemIR::FacetValue>(inst_id));
+  inst_id = context.constant_values().GetConstantInstId(inst_id);
+
+  if (auto access = context.insts().TryGetAs<SemIR::FacetAccessType>(inst_id)) {
+    inst_id = access->facet_value_inst_id;
```

Comment:
Had you considered returning directly? There's a trade-off here between single return point vs not needing to look further to determine whether the result can be used differently.
```suggestion
    return access->facet_value_inst_id;
```